aten/src/ATen/FunctionalTensorWrapper.cpp

```diff
@@ -0,0 +1,366 @@
+  outputs.reserve(t_list.size());
+  for (const auto i : c10::irange(t_list.size())) {
+    outputs[i] = unwrapFunctionalTensor(t_list[i]);
+  }
+  return outputs;
+}
+c10::List<c10::optional<Tensor>> unwrapFunctionalTensor(const c10::List<c10::optional<Tensor>> t_list) {
+  c10::List<c10::optional<Tensor>> outputs;
+  outputs.reserve(t_list.size());
+  for (const auto i : c10::irange(t_list.size())) {
+    outputs[i] = unwrapFunctionalTensor(t_list[i]);
+  }
+  return outputs;
+}
+TensorList unwrapFunctionalTensor(const TensorList& t_list) {
+    std::vector<Tensor> outputs(t_list.size());
+    for (const auto i : c10::irange(t_list.size())) {
+        outputs[i] = unwrapFunctionalTensor(t_list[i]);
+    }
+    return outputs;
+}
+
+void sync(const Tensor& t) {
+  if (t.unsafeGetTensorImpl()->is_wrapped_number()) {
+    // Unfortunately, we can't easily guarantee that wrapped numbers (scalar-tensors)
+    // get wrapped up in a FunctionalTensorWrapper object, since they skip the dispatcher.
+    // That shouldn't matter, since I don't think we're allowed to assign to wrapped numbers anyway.
+    return;
+  }
+  // Not every tensor that hits a functionalization kernel is necessarily a functional tensor.
+  // For example, xla_tensor.copy_(cpu_tensor) needs to hit the functionalization kernel
+  // to sync xla_tensor, but not cpu_tensor.
+  if (!at::functionalization::impl::isFunctionalTensor(t)) {
+      return;
+  }
+  auto functional_impl = at::functionalization::impl::unsafeGetFunctionalWrapper(t);
+  functional_impl->sync_();
+}
+void sync(const c10::optional<Tensor>& t) {
+  if (t.has_value()) {
+    sync(*t);
+  }
+}
+void sync(const c10::List<Tensor> t_list) {
+  for (const auto i : c10::irange(t_list.size())) {
+    sync(t_list[i]);
+  }
+}
+void sync(const at::TensorList t_list) {
+  for (auto t: t_list) {
+    sync(t);
+  }
+}
+void sync(const c10::List<c10::optional<Tensor>> t_list) {
+  for (const auto i : c10::irange(t_list.size())) {
+    sync(t_list[i]);
+  }
+}
+
+void maybe_add_update(Tensor& self) {
+  TORCH_INTERNAL_ASSERT(at::functionalization::impl::isFunctionalTensor(self));
+  auto functional_base_impl = at::functionalization::impl::unsafeGetFunctionalWrapper(self);
+  functional_base_impl->maybe_add_update();
+}
+
+void set_view_meta(const at::Tensor& out, const at::Tensor& t, functionalization::ViewMeta meta, int64_t out_idx) {
+  TORCH_INTERNAL_ASSERT(at::functionalization::impl::isFunctionalTensor(out));
+  auto out_impl = at::functionalization::impl::unsafeGetFunctionalWrapper(out);
+  if (out_idx != 0) {
+      // Note [out_idx in ViewMeta]
+      // When a view op outputs multiple tensors, each output needs its own separate ViewMeta.
+      // Each ViewMeta also tracks the index of the particular output tensor, which is needed in the reverse function.
+      out_impl->set_view_meta(t, functionalization::ViewMeta(meta.forward_fn, meta.reverse_fn, out_idx));
+  } else {
+      out_impl->set_view_meta(t, meta);
+  }
```

Comment:
It's not major, but I think there's an issue here - the logic rests on the assumption that the `out_index` of the passed in `ViewMeta` will be `0`, which is (if I've read far enough) just a private unchecked contract between the code-generated callers of the `set_view_meta()` overloads, those implementations, and this helper, via the default constructor for `ViewMeta`. 

A cleanish alternative that would check the assumption is a method like `meta.toOutIndex(out_idx)` that's a no-op if `out_idx == this->output_index` and makes a copy otherwise - then you could just call that here unconditionally. But at a minimum I'd at least add `|| meta.out_index != 0` to the if condition.

Reply:
a `to_out_index` method that hides that logic sounds good to me (definitely cleaner).

Also, I'm making another change that I noticed while going through the existing feedback: there's really no reason for the `set_view_meta` method. It should really just be a constructor that sets the `view_metas_` field at constructor time, instead of modifying it after-the-fact.

So `FunctionalTensorWrapper` is going to have two constructors.
- `explicit FunctionalTensorWrapper(Tensor value)` (generates a wrapper from an inner tensor)
- explicit FunctionalTensorWrapper(Tensor view_value, const FunctionalTensorWrapper* base, functionalization::ViewMeta meta)` (generates a wrapper given an inner tensor that was created off of a base tensor `base`, with `meta` describing the view operator).